csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,457 @@
+                                      : merged_pad_width);
+    }
+  }
+
+  // construct TensorDomain for output TV.
+  std::vector<IterDomain*> merged_root_ids;
+  std::vector<IterDomain*> merged_logical_ids;
+  for (const auto i : c10::irange(rank)) {
+    Val* left_pad = merged_pad_widths.at(i * 2);
+    Val* right_pad = merged_pad_widths.at(i * 2 + 1);
+    IterDomain* inp_id = inp_dom.at(i);
+    if (left_pad->isZeroInt() && right_pad->isZeroInt()) {
+      merged_root_ids.push_back(inp_id->cloneWithoutRFactor());
+      merged_logical_ids.push_back(merged_root_ids.back());
+      continue;
+    }
+    // NOTE: nvfuser pad doesn't support negative padding, so we don't have to
+    // worry about it cancelling out.
+    IterDomain* merged_root_id =
+        IterDomainBuilder(inp_id).is_rfactor_domain(true).build();
+    merged_root_ids.push_back(merged_root_id);
+    merged_logical_ids.push_back(IterDomain::resize(
+        merged_root_id,
+        left_pad,
+        right_pad,
+        true,
+        ref_iter_type.at(i)->getIterType()));
+  }
+
+  auto* new_out = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          merged_root_ids, merged_logical_ids, merged_logical_ids),
+      pad_tv->getDataType().value());
+  IrBuilder::create<PadOp>(new_out, pad_tv, merged_pad_widths, pad_value);
+  return new_out;
+}
+
+// Note [ PadOp Propagation Rule ]
+//
+// Push PadOp to its producer to reduce segmentation caused by the resize node
+// introduced in PadOp. The hope is that we can either push PadOp to inputs to
+// the fusion, or far enough that the fusion segments before the PadOp would
+// become a trivial no-op.
+//
+// The concept is that the following program:
+//   tv1 = UnaryOp(tv0)
+//   tv2 = PadOp(tv1)
+// can be replayed as the program below:
+//   tv0_padded = PadOp(tv0)
+//   tv2_new = UnaryOp(tv0_padded)
+// Given certain constraints on the UnaryOp, we can ensure that the
+// computational output remains the same when we replace all uses of `tv2` with
+// `tv2_new`.
+//
+// This function only propagates `zero-padding`, i.e. PadOp with pad value as
+// constant zero. Based on which we have certain operations that can allow PadOp
+// to propagated across. See `padCompatibleUnaryOp` and `padCompatibleBinaryOp`.
+Val* propagatePadToProducer(PadOp* pad_op) {
+  // establish pad dependencies, ensures that pad_value and pad_widths are live
+  // at the time of replay.
+  std::vector<Val*> pad_dependencies;
+  for (Val* val : pad_op->inputs()) {
+    if (val == pad_op->in() || val->isConst()) {
+      continue;
+    }
+    pad_dependencies.push_back(val);
+  }
+  auto pad_replay_check = [&pad_dependencies](Val* val) {
+    if (!val->isA<TensorView>()) {
+      return false;
+    }
+    if (std::any_of(
+            pad_dependencies.begin(),
+            pad_dependencies.end(),
+            [val](Val* pad_dependency) {
+              return DependencyCheck::isDependencyOf(pad_dependency, val);
```

Comment:
You could also use `DependencyCheck::getAllDependentVals` for `pad_dependencies` (have to make it an `unordered_set`). That way you can just check inclusion in the resulting set for each `val`.